zbus/src/address/transport/unix.rs

```diff
@@ -1,129 +1,112 @@
-            {
-                use std::os::unix::ffi::OsStrExt;
+impl<'a> TryFrom<&'a Address<'a>> for Unix<'a> {
+    type Error = Error;
 
-                encode_percents(f, path.as_bytes())?;
+    fn try_from(s: &'a Address<'a>) -> Result<Self> {
+        let mut kind = None;
+        let mut iter = s.key_val_iter();
+        for (k, v) in &mut iter {
+            match k {
+                "path" | "dir" | "tmpdir" => {
+                    let v = v.ok_or_else(|| Error::MissingValue(k.into()))?;
+                    let v = decode_percents_os_str(v)?;
+                    kind = Some(match k {
+                        "path" => UnixAddrKind::Path(v),
+                        "dir" => UnixAddrKind::Dir(v),
+                        "tmpdir" => UnixAddrKind::Tmpdir(v),
+                        // can't happen, we matched those earlier
+                        _ => panic!(),
+                    });
+                    break;
+                }
+                "abstract" => {
+                    let v = v.ok_or_else(|| Error::MissingValue(k.into()))?;
+                    let v = decode_percents(v)?;
+                    kind = Some(UnixAddrKind::Abstract(v));
+                    break;
+                }
+                "runtime" => {
+                    let v = v.ok_or_else(|| Error::MissingValue(k.into()))?;
+                    let v = decode_percents_str(v)?;
+                    if v != "yes" {
+                        return Err(Error::InvalidValue(k.into()));
+                    }
+                    kind = Some(UnixAddrKind::Runtime);
+                    break;
+                }
+                _ => continue,
             }
-
-            #[cfg(windows)]
-            write!(f, "{}", path.to_str().ok_or(std::fmt::Error)?)?;
-
-            Ok(())
         }
-
-        match self {
-            UnixSocket::File(path) => {
-                f.write_str("path=")?;
-                fmt_unix_path(f, path.as_os_str())?;
-            }
-            #[cfg(target_os = "linux")]
-            UnixSocket::Abstract(name) => {
-                f.write_str("abstract=")?;
-                fmt_unix_path(f, name)?;
-            }
-            UnixSocket::Dir(path) => {
-                f.write_str("dir=")?;
-                fmt_unix_path(f, path.as_os_str())?;
-            }
-            UnixSocket::TmpDir(path) => {
-                f.write_str("tmpdir=")?;
-                fmt_unix_path(f, path.as_os_str())?;
+        let Some(kind) = kind else {
+            return Err(Error::Other(
+                "invalid `unix:` address, missing required key".into(),
+            ));
+        };
+        for (k, _) in iter {
+            match k {
+                "path" | "dir" | "tmpdir" | "abstract" | "runtime" => {
+                    return Err(Error::Other("invalid address, only one of `path` `dir` `tmpdir` `abstract` or `runtime` expected".into()));
+                }
+                _ => (),
             }
         }
+        Ok(Unix { kind })
```

Comment:
empty line before this as well. Won't point out all of them here. Please try your best.